code/stl/Complex.hpp

```diff
@@ -3,72 +3,101 @@
 template <typename T=float>
 class Complex_t {
 public:
-    Complex_t(T r, T i);
-    Complex_t();
-
-    T real() const {return m_r;};
-    T imaginary() const {return m_i;};
-
-    Complex_t operator+(const Complex_t& other) const;
-    Complex_t operator-(const Complex_t& other) const;
-    Complex_t operator*(const Complex_t& other) const;
-    Complex_t operator*(const T factor) const;
-    Complex_t operator/(const T dividend) const;
-    Complex_t& operator+=(const Complex_t& other);
-    bool operator<(const Complex_t& a) const;
+    Complex_t() = default;
+    Complex_t(T r, T i) : m_r(r), m_i(i) {}
+
+    T real() const { return m_r; }
+    T imaginary() const { return m_i; }
+
+    T norm() const {
+        return m_r * m_r + m_i * m_i;
+    }
+
+    Complex_t& operator+=(const Complex_t& other) {
+        m_r += other.m_r;
+        m_i += other.m_i;
+        return *this;
+    }
+
+    Complex_t& operator-=(const Complex_t& other) {
+        m_r -= other.m_r;
+        m_i -= other.m_i;
+        return *this;
+    }
+
+    Complex_t& operator*=(const Complex_t& other) {
+        const auto r = m_r * other.m_r - m_i * other.m_i;
+        const auto i = m_r * other.m_i + m_i * other.m_r;
+        m_r = r;
+        m_i = i;
+        return *this;
+    }
+
+    Complex_t& operator*=(T factor) {
+        m_r *= factor;
+        m_i *= factor;
+        return *this;
+    }
+
+    Complex_t& operator/=(const Complex_t& other) {
+        const T n = other.norm();
+        m_i = (m_i * other.m_r - m_r * other.m_i) / n;
+        m_r = (m_r * other.m_r + m_i * other.m_i) / n;
+        return *this;
+    }
+
+    Complex_t& operator/=(T divisor) {
+        m_r /= divisor;
+        m_i /= divisor;
+        return *this;
+    }
+
+    friend Complex_t operator+(Complex_t a, const Complex_t& b) {
+        return a += b;
+    }
+
+    friend Complex_t operator-(Complex_t a, const Complex_t& b) {
+        return a -= b;
+    }
+
+    friend Complex_t operator*(Complex_t a, const Complex_t& b) {
+        return a *= b;
+    }
+
+    friend Complex_t operator*(Complex_t c, T factor) {
+        return c *= factor;
+    }
```

Comment:
Template to kill two birds with one stone, or too complicated?

Reply:
Three, actually. The below could also be covered.

Reply:
The template would for sure work, but yes, it would complicate the situation a bit, since errors would then be reported one template instantiation deeper. I slightly prefer to leave the code as is.